File: Game-1-modular/entities/components/inventory.py
"""Inventory system components"""

from dataclasses import dataclass
from typing import List, Optional, Dict, Any

from data.models import EquipmentItem, MaterialDefinition
from data.databases import MaterialDatabase, EquipmentDatabase


@dataclass
class ItemStack:
    item_id: str
    quantity: int
    max_stack: int = 99
    equipment_data: Optional['EquipmentItem'] = None  # For equipment items, store actual instance
    rarity: str = 'common'  # Rarity for materials and crafted items
    crafted_stats: Optional[Dict[str, Any]] = None  # Stats from minigame crafting with rarity modifiers

    def __post_init__(self):
        mat_db = MaterialDatabase.get_instance()
        if mat_db.loaded:
            mat = mat_db.get_material(self.item_id)
            if mat:
                self.max_stack = mat.max_stack

        # Equipment items don't stack
        equip_db = EquipmentDatabase.get_instance()
        is_equip = equip_db.is_equipment(self.item_id)

        if is_equip:
            self.max_stack = 1
            # Create equipment instance if not already set
            if self.equipment_data is None:
                self.equipment_data = equip_db.create_equipment_from_id(self.item_id)

    def can_add(self, amount: int) -> bool:
        return self.quantity + amount <= self.max_stack

    def add(self, amount: int) -> int:
        space = self.max_stack - self.quantity
        added = min(space, amount)
        self.quantity += added
        return amount - added

    def get_material(self) -> Optional[MaterialDefinition]:
        return MaterialDatabase.get_instance().get_material(self.item_id)

    def is_equipment(self) -> bool:
        """Check if this item stack is equipment"""
        # If equipment_data is set, it's equipment (handles invented items)
        if self.equipment_data is not None:
            return True
        return EquipmentDatabase.get_instance().is_equipment(self.item_id)

    def get_equipment(self) -> Optional[EquipmentItem]:
        """Get equipment data if this is equipment"""
        # Return the stored equipment instance if available (handles invented items)
        if self.equipment_data:
            return self.equipment_data
        # Check if this is a known equipment type from database
        if not EquipmentDatabase.get_instance().is_equipment(self.item_id):
            return None
        # Create a new one from database (shouldn't happen for equipment items with proper data)
        return EquipmentDatabase.get_instance().create_equipment_from_id(self.item_id)

    def can_stack_with(self, other: 'ItemStack') -> bool:
        """
        Check if this item can stack with another item.

        Items can stack if:
        1. Same item_id
        2. Both are not equipment (equipment never stacks)
        3. Both have identical rarity
        4. Both have identical crafted_stats (or both have None/empty stats)

        Args:
            other: Another ItemStack to check compatibility with

        Returns:
            bool: True if items can stack together
        """
        # Must be same item
        if self.item_id != other.item_id:
            return False

        # Equipment never stacks
        if self.is_equipment() or other.is_equipment():
            return False

        # Must have same rarity
        if self.rarity != other.rarity:
            return False

        # Check crafted_stats compatibility
        # Both None or empty = can stack
        self_stats = self.crafted_stats
        other_stats = other.crafted_stats

        # Normalize None and empty dict to be equivalent
        if not self_stats:
            self_stats = {}
        if not other_stats:
            other_stats = {}

        # Must have identical stats to stack
        return self_stats == other_stats


class Inventory:
    def __init__(self, max_slots: int = 30):
        self.slots: List[Optional[ItemStack]] = [None] * max_slots
        self.max_slots = max_slots
        self.dragging_slot: Optional[int] = None
        self.dragging_stack: Optional[ItemStack] = None
        self.dragging_from_equipment: bool = False  # Track if dragging from equipment slot

    def add_item(self, item_id: str, quantity: int, equipment_instance: Optional['EquipmentItem'] = None,
                 rarity: str = 'common', crafted_stats: Optional[Dict[str, Any]] = None) -> bool:
        remaining = quantity
        mat_db = MaterialDatabase.get_instance()
        equip_db = EquipmentDatabase.get_instance()

        # Equipment doesn't stack
        is_equip = equip_db.is_equipment(item_id)

        if is_equip:
            for i in range(quantity):
                empty = self.get_empty_slot()
                if empty is None:
                    return False
                # Use provided equipment instance or create new one
                equip_data = equipment_instance if equipment_instance else equip_db.create_equipment_from_id(item_id)
                if not equip_data:
                    print(f"WARNING: Could not create equipment data for {item_id}")
                    return False

                stack = ItemStack(item_id, 1, 1, equip_data, rarity=rarity, crafted_stats=crafted_stats)
                self.slots[empty] = stack
            return True

        # Normal materials can stack (but only if crafted_stats match)
        mat = mat_db.get_material(item_id)
        max_stack = mat.max_stack if mat else 99

        # Create a temporary ItemStack to check stacking compatibility
        temp_stack = ItemStack(item_id, 1, max_stack, rarity=rarity, crafted_stats=crafted_stats)

        # Try to add to existing stacks with matching crafted_stats
        for slot in self.slots:
            if slot and remaining > 0 and temp_stack.can_stack_with(slot):
                remaining = slot.add(remaining)

        # Create new stacks for remaining items
        while remaining > 0:
            empty = self.get_empty_slot()
            if empty is None:
                return False
            stack_size = min(remaining, max_stack)
            self.slots[empty] = ItemStack(item_id, stack_size, max_stack, rarity=rarity, crafted_stats=crafted_stats)
            remaining -= stack_size
        return True

    def get_empty_slot(self) -> Optional[int]:
        for i, slot in enumerate(self.slots):
            if slot is None:
                return i
        return None

    def get_item_count(self, item_id: str) -> int:
        return sum(slot.quantity for slot in self.slots if slot and slot.item_id == item_id)

    def start_drag(self, slot_index: int):
        if 0 <= slot_index < self.max_slots and self.slots[slot_index]:
            self.dragging_slot = slot_index
            self.dragging_stack = self.slots[slot_index]
            self.slots[slot_index] = None
            self.dragging_from_equipment = False

    def end_drag(self, target_slot: int):
        if self.dragging_stack is None:
            return
        if 0 <= target_slot < self.max_slots:
            if self.slots[target_slot] is None:
                self.slots[target_slot] = self.dragging_stack
            elif self.dragging_stack.can_stack_with(self.slots[target_slot]):
                # Can stack: same item, both not equipment, and matching crafted_stats
                overflow = self.slots[target_slot].add(self.dragging_stack.quantity)
                if overflow > 0:
                    self.dragging_stack.quantity = overflow
                    self.slots[self.dragging_slot] = self.dragging_stack
            else:
                # Can't stack: swap the items
                self.slots[target_slot], self.dragging_stack = self.dragging_stack, self.slots[target_slot]
                if self.dragging_stack and self.dragging_slot is not None:
                    self.slots[self.dragging_slot] = self.dragging_stack
        else:
            if self.dragging_slot is not None:
                self.slots[self.dragging_slot] = self.dragging_stack
        self.dragging_slot = None
        self.dragging_stack = None
        self.dragging_from_equipment = False

    def cancel_drag(self):
        if self.dragging_stack and self.dragging_slot is not None and not self.dragging_from_equipment:
            self.slots[self.dragging_slot] = self.dragging_stack
        self.dragging_slot = None
        self.dragging_stack = None
        self.dragging_from_equipment = False

    def has_item(self, item_id: str, quantity: int = 1) -> bool:
        """Check if inventory has at least quantity of item_id"""
        return self.get_item_count(item_id) >= quantity

    def remove_item(self, item_id: str, quantity: int = 1) -> bool:
        """Remove quantity of item_id from inventory. Returns True if successful."""
        if not self.has_item(item_id, quantity):
            return False

        remaining = quantity
        for i, slot in enumerate(self.slots):
            if slot and slot.item_id == item_id and remaining > 0:
                if slot.quantity <= remaining:
                    remaining -= slot.quantity
                    self.slots[i] = None
                else:
                    slot.quantity -= remaining
                    remaining = 0
                    break

        return remaining == 0

File: Game-1-Godot/src/Game1.Core/Data/Inventory.cs
using System.Text.Json.Nodes;

namespace Game1.Core.Data;

/// <summary>
/// Port of entities/components/inventory.py — ItemStack + Inventory (30 slots
/// default, stacking split by rarity + crafted_stats, equipment never stacks,
/// drag/merge/swap semantics). Python resolves databases via singletons; here
/// they are injected. The ITEM_ACQUIRED bus publish becomes an event hook.
/// Pinned by conformance/goldens/db_parity/inventory_buffs.json (scenarios
/// executed through the real Python classes with the booted databases).
/// </summary>
public sealed class ItemStack
{
    private readonly EquipmentDatabase? _equipDb;

    public string ItemId { get; }
    public int Quantity { get; set; }
    public int MaxStack { get; }
    public EquipmentItem? EquipmentData { get; }
    public string Rarity { get; }
    public JsonObject? CraftedStats { get; }

    // inventory.py:19-34 (__post_init__): material max_stack lookup; equipment
    // forces max_stack 1 + materializes equipment data
    public ItemStack(MaterialDatabase? matDb, EquipmentDatabase? equipDb,
                     string itemId, int quantity, int maxStack = 99,
                     EquipmentItem? equipmentData = null,
                     string rarity = "common", JsonObject? craftedStats = null)
    {
        _equipDb = equipDb;
        ItemId = itemId;
        Quantity = quantity;
        MaxStack = maxStack;
        EquipmentData = equipmentData;
        Rarity = rarity;
        CraftedStats = craftedStats;

        if (matDb is { Loaded: true } && matDb.GetMaterial(itemId) is { } mat)
            MaxStack = (int)mat.MaxStack;
        if (equipDb is not null && equipDb.IsEquipment(itemId))
        {
            MaxStack = 1;
            EquipmentData ??= equipDb.CreateEquipmentFromId(itemId);
        }
    }

    public int Add(int amount)
    {
        var space = MaxStack - Quantity;
        var added = Math.Min(space, amount);
        Quantity += added;
        return amount - added;
    }

    public bool IsEquipment() =>
        EquipmentData is not null || (_equipDb?.IsEquipment(ItemId) ?? false);

    // inventory.py:66-106 — same id, neither equipment, same rarity,
    // crafted_stats equal (null and empty are equivalent)
    public bool CanStackWith(ItemStack other)
    {
        if (ItemId != other.ItemId) return false;
        if (IsEquipment() || other.IsEquipment()) return false;
        if (Rarity != other.Rarity) return false;
        var mine = CraftedStats is { Count: > 0 } ? CraftedStats : new JsonObject();
        var theirs = other.CraftedStats is { Count: > 0 } ? other.CraftedStats : new JsonObject();
        return JsonNode.DeepEquals(mine, theirs);
    }
}

public sealed class Inventory
{
    private readonly MaterialDatabase _matDb;
    private readonly EquipmentDatabase _equipDb;

    public List<ItemStack?> Slots { get; }
    public int MaxSlots { get; }
    public int? DraggingSlot { get; private set; }
    public ItemStack? DraggingStack { get; private set; }

    /// <summary>Debug "ghost inventory" (F1): when on, the player is treated
    /// as holding an unlimited amount of every MATERIAL — GetItemCount reports
    /// a huge count and RecipeCrafting skips consumption — without using any
    /// visible slot. Off by default; never set in conformance.</summary>
    public bool DebugInfiniteMaterials;

    /// <summary>Replaces the ITEM_ACQUIRED GameEventBus publish (inventory.py:127-138).</summary>
    public event Action<string, int, string, string>? ItemAcquired;

    public Inventory(MaterialDatabase matDb, EquipmentDatabase equipDb, int maxSlots = 30)
    {
        _matDb = matDb;
        _equipDb = equipDb;
        MaxSlots = maxSlots;
        Slots = Enumerable.Repeat<ItemStack?>(null, maxSlots).ToList();
    }

    // inventory.py:117-175
    public bool AddItem(string itemId, int quantity,
                        EquipmentItem? equipmentInstance = null,
                        string rarity = "common", JsonObject? craftedStats = null)
    {
        var isEquip = equipmentInstance is not null || _equipDb.IsEquipment(itemId);
        ItemAcquired?.Invoke(itemId, quantity, isEquip ? "equipment" : "material", rarity);

        if (isEquip)
        {
            for (var i = 0; i < quantity; i++)
            {
                var empty = GetEmptySlot();
                if (empty is null) return false;
                var equipData = equipmentInstance ?? _equipDb.CreateEquipmentFromId(itemId);
                if (equipData is null) return false;
                Slots[empty.Value] = new ItemStack(_matDb, _equipDb, itemId, 1, 1,
                    equipData, rarity, craftedStats);
            }
            return true;
        }

        var remaining = quantity;
        var maxStack = _matDb.GetMaterial(itemId) is { } mat ? (int)mat.MaxStack : 99;
        var tempStack = new ItemStack(_matDb, _equipDb, itemId, 1, maxStack,
            rarity: rarity, craftedStats: craftedStats);

        foreach (var slot in Slots)
            if (slot is not null && remaining > 0 && tempStack.CanStackWith(slot))
                remaining = slot.Add(remaining);

        while (remaining > 0)
        {
            var empty = GetEmptySlot();
            if (empty is null) return false;
            var stackSize = Math.Min(remaining, maxStack);
            Slots[empty.Value] = new ItemStack(_matDb, _equipDb, itemId, stackSize,
                maxStack, rarity: rarity, craftedStats: craftedStats);
            remaining -= stackSize;
        }
        return true;
    }

    public int? GetEmptySlot()
    {
        for (var i = 0; i < Slots.Count; i++)
            if (Slots[i] is null)
                return i;
        return null;
    }

    public int GetItemCount(string itemId)
    {
        // Ghost inventory (F1 debug): unlimited of every material, no slots
        if (DebugInfiniteMaterials && _matDb.GetMaterial(itemId) is not null)
            return 999_999;
        return Slots.Where(s => s is not null && s.ItemId == itemId).Sum(s => s!.Quantity);
    }

    public bool HasItem(string itemId, int quantity = 1) =>
        GetItemCount(itemId) >= quantity;

    // inventory.py:186-191
    public void StartDrag(int slotIndex)
    {
        if (slotIndex >= 0 && slotIndex < MaxSlots && Slots[slotIndex] is not null)
        {
            DraggingSlot = slotIndex;
            DraggingStack = Slots[slotIndex];
            Slots[slotIndex] = null;
        }
    }

    // :193-215 — empty → place; stackable → merge (overflow returns to
    // origin); else swap; out-of-range → return to origin
    public void EndDrag(int targetSlot)
    {
        if (DraggingStack is null) return;
        if (targetSlot >= 0 && targetSlot < MaxSlots)
        {
            if (Slots[targetSlot] is null)
            {
                Slots[targetSlot] = DraggingStack;
            }
            else if (DraggingStack.CanStackWith(Slots[targetSlot]!))
            {
                var overflow = Slots[targetSlot]!.Add(DraggingStack.Quantity);
                if (overflow > 0)
                {
                    DraggingStack.Quantity = overflow;
                    Slots[DraggingSlot!.Value] = DraggingStack;
                }
            }
            else
            {
                var displaced = Slots[targetSlot];
                Slots[targetSlot] = DraggingStack;
                if (displaced is not null && DraggingSlot is not null)
                    Slots[DraggingSlot.Value] = displaced;
            }
        }
        else if (DraggingSlot is not null)
        {
            Slots[DraggingSlot.Value] = DraggingStack;
        }
        DraggingSlot = null;
        DraggingStack = null;
    }

    public void CancelDrag()
    {
        if (DraggingStack is not null && DraggingSlot is not null)
            Slots[DraggingSlot.Value] = DraggingStack;
        DraggingSlot = null;
        DraggingStack = null;
    }

    // :228-244
    public bool RemoveItem(string itemId, int quantity = 1)
    {
        if (!HasItem(itemId, quantity)) return false;
        var remaining = quantity;
        for (var i = 0; i < Slots.Count; i++)
        {
            var slot = Slots[i];
            if (slot is not null && slot.ItemId == itemId && remaining > 0)
            {
                if (slot.Quantity <= remaining)
                {
                    remaining -= slot.Quantity;
                    Slots[i] = null;
                }
                else
                {
                    slot.Quantity -= remaining;
                    remaining = 0;
                    break;
                }
            }
        }
        return remaining == 0;
    }
}
